lib/Yield/Assembly/HTML.cpp

```diff
@@ -527,62 +637,120 @@ static std::string instruction(const assembly::Instruction &Instruction,
-                              const model::Binary &Binary) {
-  std::string Result;
+template<bool UseVerticalLayout>
+static std::string basicBlock(const yield::BasicBlock &BasicBlock,
+                              const yield::Function &Function,
+                              const model::Binary &Binary,
+                              bool HasLabel) {
+  revng_assert(!BasicBlock.Instructions.empty());
 
+  // Compile the list of delayed instructions so the corresponding comment
+  // can be emited.
   llvm::SmallVector<MetaAddress, 2> DelayedList;
   bool IsNextInstructionDelayed = false;
   for (const auto &Instruction : BasicBlock.Instructions) {
     if (IsNextInstructionDelayed)
       DelayedList.emplace_back(Instruction.Address);
     IsNextInstructionDelayed = Instruction.HasDelayedSlot;
   }
+  revng_assert(IsNextInstructionDelayed == false,
+               "Last instruction has a unfilled delayed slot.");
 
+  // Determine the last non-delayed instruction.
+  // This is the instruction "targets" get printed for in horizontal layout.
   MetaAddress LastNotDelayedInstruction = MetaAddress::invalid();
   for (const auto &Instruction : llvm::reverse(BasicBlock.Instructions)) {
     if (!llvm::is_contained(DelayedList, Instruction.Address)) {
       LastNotDelayedInstruction = Instruction.Address;
       break;
     }
   }
+  revng_assert(LastNotDelayedInstruction.isValid());
 
-  Result += label(BasicBlock, Metadata, Binary);
+  // String the instructions together.
+  std::string Result;
   for (const auto &Instruction : BasicBlock.Instructions) {
-    Result += instruction(Instruction,
-                          llvm::is_contained(DelayedList, Instruction.Address),
-                          LastNotDelayedInstruction == Instruction.Address,
-                          BasicBlock,
-                          Metadata,
-                          Binary);
+    bool IsInDelayedSlot = llvm::is_contained(DelayedList, Instruction.Address);
+    bool ShouldPrintTargets = LastNotDelayedInstruction == Instruction.Address;
+    Result += instruction<UseVerticalLayout>(Instruction,
+                                             BasicBlock,
+                                             Function,
+                                             Binary,
+                                             IsInDelayedSlot,
+                                             ShouldPrintTargets);
   }
 
-  if (IsNextInstructionDelayed == true) {
-    std::string DelayedError = "Error: Last instruction has a delayed slot.";
-    Result += error(BasicBlock, std::move(DelayedError), 2, true);
+  if (HasLabel)
+    return llvm::formatv(templates::SimpleDiv,
+                         tags::BasicBlock,
+                         std::move(Result));
+  else
+    return llvm::formatv(templates::BlockDiv,
+                         tags::BasicBlock,
+                         basicBlockID(BasicBlock.Address),
+                         std::move(Result));
+}
+#include <iostream>
```

Comment:
Wat

Reply:
🤦🏻, an indication of some painful debugging at one point I forgot to remove. There's little as embarrassing as for someone to see your dirty includes 😓

Reply:
As if it was never there.